src/shogun/evaluation/CrossValidation.cpp

```diff
@@ -280,16 +286,22 @@ float64_t CCrossValidation::evaluate_one_run()
 			/* set feature subset for training */
 			SGVector<index_t> inverse_subset_indices=
 					m_splitting_strategy->generate_subset_inverse(i);
-			m_features->add_subset(inverse_subset_indices);
+
+			CFeatures* features=CFeatures::create_shallow_copy(m_features);
```

Comment:
keep in mind that cloning the machine above also will clone the features (including the matrix), if the features are set.
I would make use of that and not separately clone the features here. I get it that you want to be efficient here, but let's postpone that for a bit and get the other things right first.